### ISSUE TITLE
Legacy git_pillar broken in 2016.3.1 when using environments

### ISSUE THREAD

**Issue by peter-slovak** [saltstack/salt#34962]
### Description of Issue

After an upgrade from 2015.8.5 to 2016.3.1, Git pillar in Legacy mode (with `__env__` supplied as the branch) stopped being usable. We traced the issue to the master omitting the `__opts__['environment']` minion option, which is the reason Git pillar never considers any branch different than 'gitfs_base' (master in most cases).

To be sure, we tested the same workflow in a 2015.8.5 environment (master + minions), where the "environment" option correctly arrived from minion to master and up to the code in git_pillar.py . Based on that value, `_LegacyGitPillar()` was able to determine the correct Git branch to load Pillar data from and do the processing. However, in 2016.3.1, `__opts__['environment']` is None and `_LegacyGitPillar()` always builds the pillar from "base", a.k.a. master branch.
### Setup

Git pillar config on saltmaster:

``` bash
ext_pillar:
  - git: __env__ ssh://git@git.domain.com/group/salt.git root=pillar
```

(no other git pillar-related options)
### Steps to Reproduce Issue
- use legacy Git pillar on salt master
- use environments on minions (the 'environment' config option)
- use Salt 2016.3.1
- preferably turn of pillar_cache on minions
- request pillar data, either locally or remotely
- result: the pillar always gets compiled from "base" environment 
### Versions Report

``` bash
Salt Version:
           Salt: 2016.3.1

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.3.7
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.11+ (default, Apr 17 2016, 14:00:29)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: Ubuntu 16.04 xenial
        machine: x86_64
        release: 4.4.0-24-generic
         system: Linux
        version: Ubuntu 16.04 xenial
```

**EDIT:** The issue with configuration variable "pillarenv" is the same; `__opts__['pillarenv']` is None on the master no matter the configuration on minion.

**Comment by Ch3LL** [comment]
@terminalmage would you mind taking a look at this.. Its my understanding that this should not work in 2015.8 and above but it looks like they are reporting it as working in 2015.8. Should someone expect using the legacy configuration to work in 2016.3?

**Comment by peter-slovak** [comment]
@Ch3LL Yes, at first the feature of dynamic branch mapping with `__env__` was about to be dropped, but then #28282 has been merged into 2015.8.0, so we expected the legacy git pillar to keep working. The merge was a response to #28125.

**Comment by peter-slovak** [comment]
I've looked a bit more into this - it's not as much a problem with LegacyGitPillar as it is with LazyLoader. Somehow, when pillar.git_pillar is being lazyloaded, the __opts__ dunder dict for the module is incomplete (missing the "environment" and "pillarenv" vars). Maybe it is loaded too early or in a stage that is out of context for the minion pillar request - I can't see into the loading process very well.

So, even if we implemented dynamic environment name in non-legacy pillar (which I was determined to try), it looks like it still wouldn't work because of insufficient information in the opts coming from minion.

It's important to mention that the opts dict arrives fine from the minion and "environment" makes its way to many other lazyloaded modules, but not to git_pillar.

**Comment by terminalmage** [comment]
@peter-slovak I cannot reproduce this with the legacy git_pillar syntax on the head of the 2016.3 branch. I tried with a repo that has no master branch just to be sure that there was no way that it would try to checkout ``master``. I also set environment to a different branch, one with a top file using ``{{env}}`` in place of the environment name.

This may indeed be a moot point, as I have implemented ``__env__`` support for new-style git_pillar in https://github.com/saltstack/salt/pull/36659.
